### PR TITLE
brainstorming rules for components and producers

### PR DESCRIPTION
**Pull request by alexandru-filimon**
These rules are up for debate

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Other information**:

### CODE REVIEW

File: docs/docs/best-practices.md [hunk 1]
```diff
@@ -55,3 +55,32 @@ enum Modes {
...
+...
+<Component>.producers(Object.values(producers)); 
+```
+- if you use typescript keep using the types throughout the entire producer (use type casting if needed)
+
+### Components:
+- only one component per file
+- they should have a descriptive name (PascalCase) and the file should have the same name
+- they should do only one thing
+- they should use `get` instead of `observe` if you don’t need that value to retrigger the component
```
Comment (dumconstantin) [inline docs/docs/best-practices.md:76]: I think `get` should be avoided in views. Views should be.... views of data - their responsibility is to always be up to date.

If you can think of a use case where you would need `get` in a component, please let me know, that would a be an interesting conversation!

File: docs/docs/best-practices.md [hunk 2]
```diff
@@ -55,3 +55,32 @@ enum Modes {
...
+- the use of layout components is highly encourage (cards, sidebars, navbars)
+- use as few props as posible, they should get their data from the state
```
Comment (dumconstantin) [inline docs/docs/best-practices.md:78]: Maybe we can elaborate on this point - e.g. don't pass down props that contain data that the child component could simply get from the state....

File: docs/docs/best-practices.md [hunk 3]
```diff
@@ -55,3 +55,32 @@ enum Modes {
...
+- if they have sub-components they can be put in a `./component` folder on the same level
+- they should be used only for **displaying** data or as **layout component** so they should manipulate the data as litle as posible
```
Comment (dumconstantin) [inline docs/docs/best-practices.md:80]: In regards to data manipulation, maybe we can be more explicit - e.g. don't process data in a view or something like that

File: docs/docs/best-practices.md (hunk 1 above)
Comment (alexandru-filimon) [reply docs/docs/best-practices.md:76]: Maybe if you have some semi-permanent values in the initial state for a profile/theme for exemple

File: docs/docs/best-practices.md (hunk 2 above)
Comment (alexandru-filimon) [reply docs/docs/best-practices.md:78]: Agree

File: docs/docs/best-practices.md (hunk 3 above)
Comment (alexandru-filimon) [reply docs/docs/best-practices.md:80]: I agree

**Comment by dumconstantin** [comment]
Convention for naming variables: 

- `observe`: the final name in the path or an interpretation of that value:
`bam = observe.something.bam`
`title = observe.article.title`
`isFinished = observe.article.finishedWritingInArticle`

- `update`: the prefix `update` + the final name in the path or an interpretation of that value:
`updateBam = update.something.bam`
`updateTitle = update.article.title`
`updateIsFinished = update.article.finishedWritingInArticle`

- `get`: the prefix `get` + the final name in the path or an interpretation of that value:
`getBam = get.something.bam`
`getTitle = get.article.title`
`getIsFinished = get.article.finishedWritingInArticle`

- `_` for pseudo-private props:
`_viewId`
`_producerId`
`_now`

Example:
```
const A: producer = ({
  foo = observe.foo,
  updateFoo = update.foo,
  getFoo = get.foo
}) => {
  ...
}
```

File: docs/docs/best-practices.md [hunk 4]
```diff
@@ -55,3 +55,32 @@ enum Modes {
...
+### Producers:
+- only one producer per file
+- they should have a descriptive name (camelCase) and the file should have the same name
+- they should do only one thing
+- they should use `get` instead of `observe` if you don’t need that value to retrigger the producer
+- in the folder structure they should stay close to the components that use them. 
+- if there are more than one producer for a component the propper way of **importing** them should be to by creating an `index.js/ts` file in the producers folder and export them all there and in the component write like this
+``` import * as producers from "./producers";
+...
+<Component>.producers(Object.values(producers));
```
Comment (dumconstantin) [inline docs/docs/best-practices.md:68]: `Component.producers(producers)`
The fn accepts a producer, an object or an array of producers

File: docs/docs/best-practices.md (hunk 1 above)
Comment (dumconstantin) [reply docs/docs/best-practices.md:76]: Using `get` should be reserved to events that happen within the view e.g. `onClick`.
```
const A: view = ({
  foo= observe.foo,
  updateBar = update.bar,
  getId = get.selectedId
}) => {
  return <div onClick={() => updateBar.set(getId.value()) }>CLick me</div>
}
```

**Pull request merged**